### PR TITLE
fix(ws): turn-in-flight signal + stream-token message_id for reconnect resync (#1164)

### LINKED ISSUE #1164: WS reconnect: replay/resync protocol for an NPC turn in flight across a disconnect
## Background

PR #1158 added the first WebSocket reconnect resync for the web frontend: on
reconnect, `+page.svelte` re-fetches snapshot/map/npcs, `StreamManager.reset()`
discards orphaned pending turns and finalizes half-streamed `textLog` entries,
and `onStreamToken` re-asserts `streamingActive` so a resumed stream keeps input
disabled. This strictly improves on the prior behaviour (the socket silently
reconnected and dropped every event in the gap).

Two narrow, browser-only **P2** edge cases remain that can't be resolved cleanly
on the client alone — they need a small server-side contribution. Deferred from
#1158 by decision to land the audit fixes.

## Remaining edge cases (from Codex review on #1158)

1. **Pre-token busy gap** ([thread](https://github.com/dmooney/Rundale/pull/1158#discussion_r3331281803))
   When the socket reconnects while the backend is still processing a turn but
   has not yet emitted the next `stream-token` (slow model thinking, or a long
   pause between batches), the reconnect handler clears `streamingActive`, so the
   editor/travel chips become usable while the original `submit_input` is still
   in flight — a duplicate-turn window until the next token arrives and
   re-asserts the flag. (Quick-travel has no `isSubmitting` guard.)

2. **Lost placeholder id on resumed stream** ([thread](https://github.com/dmooney/Rundale/pull/1158#discussion_r3331281806))
   If the socket drops after the backend emits the empty `text-log` placeholder
   (which carries the message id) but before the first `stream-token`,
   `StreamManager.reset()` removes the only client copy of that turn's id. When
   the backend resumes tokens after reconnect, `StreamTokenPayload` carries no
   message id, so the rebuilt bubble has no `entry.id` → the completed NPC
   message is non-reactable and its language hints aren't keyed to it.

## Root cause

There is no replay/resync contract between server and client for an NPC turn
that is in flight across a reconnect. The client cannot distinguish "stream is
dead" from "stream is slow/paused-but-alive", and a resumed stream has no id.

## Proposed direction

- Add a server-authoritative **turn-in-flight** signal the client can read on
  reconnect (e.g. surface it in `WorldSnapshot` / the snapshot the resync
  fetches), so `streamingActive` can be set from authoritative state instead of
  the client guessing.
- Add `message_id` (and `stream_turn_id`) to `StreamTokenPayload` so a resumed
  stream rebinds to the correct, reactable `textLog` entry and its hints.
- Consider a lightweight replay/sequence mechanism on `/api/ws` so events
  emitted during the disconnect gap can be re-delivered rather than dropped.

This is a cross-cutting change (parish-core IPC types + parish-server WS +
frontend), so it's tracked separately from the frontend audit PR (#1158).

### PR DESCRIPTION
**Pull request by dmooney**
Closes #1164.

## Background

PR #1158 added the first WebSocket reconnect resync, but two browser-only **P2** edge cases remained because there was no server↔client contract for an NPC turn that is *in flight across a reconnect* — the client had to guess "dead" vs "slow-but-alive", and a resumed stream had no id.

## What this does

Two server-authoritative additions to the IPC contract, plus the frontend wiring that consumes them:

- **`WorldSnapshot.turn_in_flight`** — set by `GET /api/world-snapshot` (the exact endpoint the reconnect resync re-fetches) from the authoritative `ConversationRuntimeState::conversation_in_progress`. On reconnect the frontend re-asserts `streamingActive` from it, closing the **pre-token busy gap**: the input field and quick-travel chips stay disabled while a turn is mid-flight, eliminating the duplicate-turn window.
- **`StreamTokenPayload.message_id`** — carried for player-initiated NPC conversation turns. A stream that **resumes after a reconnect** rebinds to the reactable `textLog` entry even though `StreamManager.reset()` discarded the client's only copy of the placeholder id, so the completed NPC message stays reactable and its language hints stay keyed to it.

`message_id` is `None` for arrival-reaction streams (movement / Tauri) — those have no reconnect-resume contract and the wire shape is unchanged there (`skip_serializing_if`). The Tauri desktop transport never reconnects; the field is propagated there for type parity only.

### Out of scope

The issue's optional "Consider" bullet — a lightweight WS event replay/sequence buffer — is intentionally **not** included. The two concrete signals above close both reported edge cases without it.

## Behaviour changes & tests

New tests for every behaviour change:
- `parish-server`: `world_snapshot_reports_turn_in_flight_from_conversation_state`, `stream_tokens_carry_the_placeholder_message_id` (drives the real orchestration; asserts every `stream-token`'s `message_id` equals the placeholder id).
- frontend (`stream-manager.test.ts`): 3 resumed-stream rebind tests + 3 reconnect re-assert tests.

## Commands run

- `cargo test -p parish-core -p parish-server` — green
- `cargo build -p parish-tauri` — green
- `cargo clippy -p parish-core -p parish-server -p parish-tauri` — no warnings
- `npx vitest run` — 38 files / 462 tests green (`stream-manager` 20/20)
- `bash parish/scripts/agent-check.sh` — passed

## Live proof

A real `parish-server` process (`PARISH_PROVIDER=simulator`) was driven with live player input producing a streamed NPC turn (proof bundle attached to this PR):

- **AC1:** `GET /api/world-snapshot` returned `turn_in_flight=true` for the full ~600 ms streaming window and `false` when idle.
- **AC2:** all 24 `stream-token` frames for the turn carried `message_id='msg-6'`, identical to the `text-log` placeholder's `id='msg-6'` → MATCH.

🤖 https://claude.ai/code/session_01Ef2rMeuF1m32sPNpKZBwWN

---
_Generated by [Claude Code](https://claude.ai/code/session_01Ef2rMeuF1m32sPNpKZBwWN)_

### CODE REVIEW

**Comment by dmooney** [comment]
## Proof: 1164

### Acceptance criteria

# Acceptance criteria — #1164: WS reconnect replay/resync for an NPC turn in flight

## Background

PR #1158 added the first WS reconnect resync. Two browser-only P2 edge cases
remained because there was no server↔client contract for an NPC turn that is
*in flight across a reconnect*:

1. **Pre-token busy gap** — on reconnect the client clears `streamingActive`
   even though the backend is still mid-turn (no `stream-token` emitted yet),
   re-enabling the input field and quick-travel chips → duplicate-turn window.
2. **Lost placeholder id on resumed stream** — `StreamManager.reset()` discards
   the only client copy of the turn's message id, and a resumed `stream-token`
   carried no id, so the rebuilt NPC bubble was non-reactable and its language
   hints weren't keyed to it.

## Scope

Two server-authoritative additions to the IPC contract + the frontend wiring
that consumes them. The optional "lightweight replay/sequence mechanism" from
the issue's *Consider* bullet is intentionally **out of scope** — it is a
larger change and the two concrete signals below close both reported edge
cases.

## Observable criteria

- **AC1 — turn-in-flight signal in the resync snapshot.**
  `WorldSnapshot` gains a `turn_in_flight: bool` field. `GET /api/world-snapshot`
  (the endpoint the reconnect resync re-fetches) reports `turn_in_flight: true`
  while an NPC conversation turn is being processed and `false` when the engine
  is idle. Proven by the JSON returned from the live server.

- **AC2 — `stream-token` carries the message id.**
  `StreamTokenPayload` gains `message_id`. Every `stream-token` emitted for a
  player-initiated NPC conversation turn carries the same `message_id` as the
  turn's `text-log` placeholder `id`. Proven by the live event/command stream
  showing a non-empty `message_id` equal to the placeholder id.

- **AC3 — reconnect keeps input disabled when a turn is in flight.**
  On WS reconnect, after `StreamManager.reset()`, the frontend reads the
  authoritative snapshot: when `turn_in_flight === true` it leaves
  `streamingActive` set (input field + quick-travel chip stay disabled); when
  `false` it clears `streamingActive` as before. Proven by a frontend unit test.

- **AC4 — resumed stream rebinds to a reactable id.**
  When a `stream-token` carrying `message_id` arrives for a turn that has no
  pending state (post-reset), the rebuilt `textLog` entry adopts
  `entry.id = message_id`, so the completed NPC message is reactable and its
  hints key to it. Proven by a `stream-manager` unit test.

- **AC5 — no mode-parity / regression breakage.**
  `cargo test -p parish-core -p parish-server -p parish-tauri` and
  `just ui-test` stay green; the cross-mode equivalence test still passes.

## Non-goals

- WS event replay/sequence buffer (deferred; "Consider" bullet).
- Any change to the Tauri desktop transport behaviour (it never reconnects);
  the new field is propagated there only for type/shape parity.

### Evidence

# Evidence — #1164: WS reconnect resync for an NPC turn in flight

Evidence type: live gameplay transcript

## What ran

A real `parish-server` process (`cargo run -p parish-server -- --port 3030`,
`PARISH_PROVIDER=simulator`) was driven with live player input that produced a
real streamed NPC turn. Two observers shared one session cookie:

- `.proofs/1164/ac1_turn_in_flight.sh` — fires `POST /api/command` in the
  background and polls `GET /api/world-snapshot` (the exact endpoint the
  browser reconnect resync re-fetches).
- `.proofs/1164/ac2_ws.py` — opens a raw WebSocket to `/api/ws` and captures
  the raw event frames while a separate `curl` process fires the turn.

Full output: `.proofs/1164/transcript.txt`. Commands are reproducible via the
two scripts in this directory.

## Criterion → proof

### AC1 — turn-in-flight signal in the resync snapshot

Live transcript (`AC1` section):

```
[before] turn_in_flight=False        <- idle
  t=01 turn_in_flight=True            <- turn in flight (streaming)
  ...
  t=12 turn_in_flight=True
  t=13 turn_in_flight=False           <- turn complete
  ...
[after]  turn_in_flight=False
```

`GET /api/world-snapshot` reports `turn_in_flight=true` for the full ~600 ms the
NPC turn is streaming, then `false` once it completes — exactly the
authoritative signal the reconnect handler needs to avoid re-enabling input
mid-turn. Deterministic backing: `parish-server`
`world_snapshot_reports_turn_in_flight_from_conversation_state` asserts the
endpoint reflects `ConversationRuntimeState::conversation_in_progress`.

### AC2 — `stream-token` carries the placeholder message id

Live transcript (`AC2` section):

```
[ws] 36 events: {'world-update': 4, 'text-log': 1, 'loading': 5,
                 'stream-token': 24, 'stream-turn-end': 1, 'stream-end': 1}
  placeholder text-log: stream_turn_id=6 id='msg-6'
  stream-token: turn_id=6 message_id='msg-6' token='Old mccarthy ' -> MATCH
```

Every `stream-token` for the turn carries `message_id='msg-6'`, identical to the
`text-log` placeholder's `id` — so a stream that resumes after a reconnect
rebinds to the reactable chat entry. Deterministic backing: `parish-server`
`stream_tokens_carry_the_placeholder_message_id` asserts every `stream-token`'s
`message_id` equals the placeholder id over the real orchestration.

### AC3 — reconnect keeps input disabled when a turn is in flight

The reconnect handler in `parish/apps/ui/src/routes/+page.svelte` runs
`sm.reset(); streamingActive.set(false);` then re-asserts from the authoritative
snapshot: `if (snap.turn_in_flight) streamingActive.set(true);`. AC1 proves the
field is `true` mid-turn on the live wire, so on a mid-turn reconnect the input
field and quick-travel chips stay disabled. The store-level contract (true →
stay disabled, false/absent → cleared) is locked by the three
`reconnect re-asserts streamingActive from turn_in_flight (#1164 AC3)` unit
tests in `stream-manager.test.ts`.

### AC4 — resumed stream rebinds to a reactable id

`stream-manager.test.ts` →
`createStreamManager — resumed stream rebinds to a reactable id (#1164)`:
- `queuePendingTurn` carries the resumed `message_id` and `ensureTurnEntry`
  mints the bubble with `id = msg-42` (reactable).
- `appendStreamToken` rebuilds a post-reset entry adopting `id = msg-9`.
- a redundant later id does not clobber a placeholder-set id.

### AC5 — no mode-parity / regression breakage

- `cargo test -p parish-core -p parish-server` — all green (incl. the cross-mode
  equivalence test and the two new #1164 tests).
- `cargo build -p parish-tauri` — green (literals updated for the new fields).
- `npx vitest run` — 38 files / 462+ tests green; `stream-manager` 20/20.
- `cargo clippy` on the three crates — no warnings.

## Acceptance criteria → output line map

| Criterion | Proof |
| --- | --- |
| AC1 | transcript `AC1` block (`turn_in_flight` True t=01..12, False after) + server unit test |
| AC2 | transcript `AC2` block (`message_id='msg-6'` == placeholder `id='msg-6'` MATCH) + server unit test |
| AC3 | `+page.svelte` reconnect guard + 3 AC3 store-contract unit tests; field proven live (AC1) |
| AC4 | 3 `stream-manager` rebind unit tests |
| AC5 | cargo test/build/clippy + vitest all green |

### Transcript

```text
================================================================
  #1164 LIVE PROOF — parish-server (PARISH_PROVIDER=simulator)
  process: cargo run -p parish-server -- --port 3030  (pid 3066)
  date: 2026-06-02T11:31:34Z
================================================================

########## AC1: turn_in_flight true mid-turn, false when idle ##########
[before] turn_in_flight=False
[command] POST /api/command text='hello there' (pid 5505, blocks for the turn)
  t=01 turn_in_flight=True
  t=02 turn_in_flight=True
  t=03 turn_in_flight=True
  t=04 turn_in_flight=True
  t=05 turn_in_flight=True
  t=06 turn_in_flight=True
  t=07 turn_in_flight=True
  t=08 turn_in_flight=True
  t=09 turn_in_flight=True
  t=10 turn_in_flight=True
  t=11 turn_in_flight=True
  t=12 turn_in_flight=True
  t=13 turn_in_flight=False
  t=14 turn_in_flight=False
  t=15 turn_in_flight=False
  t=16 turn_in_flight=False
  t=17 turn_in_flight=False
  t=18 turn_in_flight=False
  t=19 turn_in_flight=False
  t=20 turn_in_flight=False
  t=21 turn_in_flight=False
  t=22 turn_in_flight=False
  t=23 turn_in_flight=False
  t=24 turn_in_flight=False
  t=25 turn_in_flight=False
  t=26 turn_in_flight=False
  t=27 turn_in_flight=False
  t=28 turn_in_flight=False
  t=29 turn_in_flight=False
  t=30 turn_in_flight=False
[after]  turn_in_flight=False

########## AC2: stream-token.message_id == placeholder text-log id ##########
[session] cookie: parish_sid=jgwthXDYAMXOnUBDm4Hhgw
[ws] connected (HTTP 101)
[command] curl POST /api/command  text='hello there' (separate process)
[ws] 36 events: {'world-update': 4, 'text-log': 1, 'loading': 5, 'stream-token': 24, 'stream-turn-end': 1, 'stream-end': 1}
  placeholder text-log: stream_turn_id=6 id='msg-6'
  stream-token: turn_id=6 message_id='msg-6' token='Old mccarthy ' -> MATCH
```

### Judge

# Judge verdict — #1164

## Scope reviewed

Server-authoritative reconnect contract for an NPC turn in flight: a
`turn_in_flight` field on `WorldSnapshot` (set by `GET /api/world-snapshot`
from `conversation_in_progress`) and a `message_id` on `StreamTokenPayload`
(populated for player-initiated NPC turns), plus the frontend wiring that
consumes both on WS reconnect.

## Criterion-by-criterion

- **AC1 (turn-in-flight signal): met.** Live transcript shows
  `GET /api/world-snapshot` returning `turn_in_flight=true` for the entire
  ~600 ms streaming window of a real NPC turn and `false` when idle, on the
  exact endpoint the reconnect resync re-fetches. Backed deterministically by
  `world_snapshot_reports_turn_in_flight_from_conversation_state`.

- **AC2 (stream-token message_id): met.** Live WS capture shows all 24
  `stream-token` frames for the turn carrying `message_id='msg-6'`, identical to
  the `text-log` placeholder `id='msg-6'`. Backed by
  `stream_tokens_carry_the_placeholder_message_id`.

- **AC3 (reconnect keeps input disabled): met.** The `+page.svelte` reconnect
  handler re-asserts `streamingActive` from `snap.turn_in_flight`; AC1 proves
  the field is true mid-turn; three store-contract unit tests lock the
  true/false/absent decision.

- **AC4 (resumed stream rebinds to reactable id): met.** Three `stream-manager`
  unit tests prove the resumed `message_id` is adopted as `entry.id` whether the
  placeholder was dropped or rebuilt, and is not clobbered when already set.

- **AC5 (no regressions): met.** Full Rust + UI suites and clippy green; the
  optional WS replay buffer was correctly left out of scope per the issue.

## Risk / debt notes

- `message_id` is `None` for arrival-reaction streams (movement, Tauri); this is
  intentional and documented — those streams have no reconnect-resume contract.
  The wire shape is unchanged there (`skip_serializing_if`).
- The Tauri desktop transport never reconnects; the new field is propagated for
  type parity only. No behaviour change there.
- No new `#[allow]`, no unexplained suppressions, no architecture-fitness
  violation (changes stay in `parish-core` IPC + entry-point wiring).

Verdict: sufficient
Technical debt: clear
Acceptance criteria: met

### Artifacts

- `transcript.txt` (full transcript)




---
_Generated by [Claude Code](https://claude.ai/code)_

File: parish/crates/parish-server/src/routes.rs [hunk 1]
```diff
@@ -57,6 +57,10 @@ pub async fn get_world_snapshot(Extension(state): Extension<Arc<AppState>>) -> J
     let mut snapshot = parish_core::ipc::snapshot_from_world(&world);
     snapshot.name_hints =
         parish_core::ipc::compute_name_hints(&world, &npc_manager, &state.pronunciations);
+    // Surface whether an NPC turn is in flight so the web frontend can
+    // re-assert `streamingActive` from authoritative state after a WebSocket
+    // reconnect, instead of guessing and re-enabling input mid-turn (#1164).
+    snapshot.turn_in_flight = state.conversation.lock().await.conversation_in_progress;
```
[the remark this answers is not in the document]
Comment (dmooney) [reply parish/crates/parish-server/src/routes.rs:69]: Good catch — applied in 940b7ea. The `world` + `npc_manager` locks are now scoped to a block and released before the `conversation` lock is awaited, so this hot reconnect-path endpoint never holds three locks at once.

---
_Generated by [Claude Code](https://claude.ai/code)_
